compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,218 @@
+                name, args, module, ..
+            } => {
+                let key = (module.clone(), name.clone());
+
+                if let Some(typ_alias) = self.type_aliases.get(&key) {
+                    // Type has been aliased. eg: import mod1.{type Cat as c} or type UserId = Int
+                    if self.name_clashes_if_unqualified(typ_alias, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(typ_alias.as_str());
+                } else if module == PRELUDE_MODULE_NAME || *module == self.current_module {
+                    if self.name_clashes_if_unqualified(name, module) {
+                        typ_str.push_str(module);
+                        typ_str.push('.');
+                    }
+                    typ_str.push_str(name.as_str());
+                } else {
+                    // Always qualify types from other modules
+                    typ_str.push_str(self.module_aliases.get(module).unwrap_or(module).as_str());
+                    typ_str.push('.');
+                    typ_str.push_str(name.as_str());
+                }
+
+                if !args.is_empty() {
+                    typ_str.push('(');
+                    typ_str.push_str(&self.args_to_string(args));
+                    typ_str.push(')');
+                }
+                let _ = self
+                    .printed_types
+                    .insert(EcoString::from(typ_str.clone()), module.clone());
+                typ_str
+            }
+            Type::Fn { args, retrn } => {
+                typ_str.push_str("fn(");
+                typ_str.push_str(&self.args_to_string(args));
+                typ_str.push_str(") -> ");
+                typ_str.push_str(&self.print(retrn));
+
+                typ_str
+            }
+            Type::Var { type_: typ, .. } => match *typ.borrow() {
+                TypeVar::Link { type_: ref typ, .. } => self.print(typ),
+                TypeVar::Unbound { id, .. } | TypeVar::Generic { id, .. } => {
+                    if let Some(crate::ast::TypeAst::Var(type_var)) = self.type_parameters.get(&id)
+                    {
+                        return type_var.name.to_string();
+                    }
+                    self.generic_type_var(id)
+                }
+            },
+            Type::Tuple { elems, .. } => {
+                typ_str.push_str("#(");
+                typ_str.push_str(&self.args_to_string(elems));
+                typ_str.push(')');
+
+                typ_str
+            }
+        }
+    }
+
+    fn name_clashes_if_unqualified(&mut self, type_: &EcoString, module: &str) -> bool {
+        match self.printed_types.get(type_) {
+            None => false,
+            Some(previous_module) if module == previous_module => false,
+            Some(_different_module) => true,
+        }
+    }
+
+    fn args_to_string(&mut self, args: &[Arc<Type>]) -> String {
```

Comment:
Rather than build an extra string all methods should push to the existing string buffer